From ac890066375f75ebdb29fd9fd900ac553777180e Mon Sep 17 00:00:00 2001
From: Connor Abbott <cwabbott0@gmail.com>
Date: Wed, 26 Oct 2022 13:05:23 +0200
Subject: [PATCH 1/2] freedreno/fdl: Expose UBWC height

We will need this for clears operating directly on UBWC.
---
 src/freedreno/fdl/fd6_layout.c       | 6 +++---
 src/freedreno/fdl/freedreno_layout.h | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/freedreno/fdl/fd6_layout.c b/src/freedreno/fdl/fd6_layout.c
index dd09a3c6d546..5dca2898b56c 100644
--- a/src/freedreno/fdl/fd6_layout.c
+++ b/src/freedreno/fdl/fd6_layout.c
@@ -185,8 +185,8 @@ fdl6_layout(struct fdl_layout *layout, enum pipe_format format,
    }
    layout->ubwc_width0 = align(DIV_ROUND_UP(ubwc_width0, ubwc_blockwidth),
                                RGB_TILE_WIDTH_ALIGNMENT);
-   ubwc_height0 = align(DIV_ROUND_UP(ubwc_height0, ubwc_blockheight),
-                        ubwc_tile_height_alignment);
+   layout->ubwc_height0 = align(DIV_ROUND_UP(ubwc_height0, ubwc_blockheight),
+                                ubwc_tile_height_alignment);
 
    uint32_t min_3d_layer_size = 0;
 
@@ -257,7 +257,7 @@ fdl6_layout(struct fdl_layout *layout, enum pipe_format format,
 
          uint32_t meta_pitch = fdl_ubwc_pitch(layout, level);
          uint32_t meta_height =
-            align(u_minify(ubwc_height0, level), ubwc_tile_height_alignment);
+            align(u_minify(layout->ubwc_height0, level), ubwc_tile_height_alignment);
 
          ubwc_slice->size0 =
             align(meta_pitch * meta_height, UBWC_PLANE_SIZE_ALIGNMENT);
diff --git a/src/freedreno/fdl/freedreno_layout.h b/src/freedreno/fdl/freedreno_layout.h
index 52be661d29eb..7e0311bd59d4 100644
--- a/src/freedreno/fdl/freedreno_layout.h
+++ b/src/freedreno/fdl/freedreno_layout.h
@@ -97,7 +97,7 @@ struct fdl_layout {
    struct fdl_slice slices[FDL_MAX_MIP_LEVELS];
    struct fdl_slice ubwc_slices[FDL_MAX_MIP_LEVELS];
    uint32_t pitch0;
-   uint32_t ubwc_width0;
+   uint32_t ubwc_width0, ubwc_height0;
    uint32_t layer_size;
    uint32_t ubwc_layer_size; /* in bytes */
    bool ubwc : 1;
-- 
GitLab


From e87de9bc5cc0483de49a13e5a221b8b7c5ce75a2 Mon Sep 17 00:00:00 2001
From: Connor Abbott <cwabbott0@gmail.com>
Date: Mon, 24 Oct 2022 13:34:21 +0200
Subject: [PATCH 2/2] tu: Clear UBWC flags buffer on transition from UNDEFINED

This fixes a hang on a650 when running these two tests back-to-back on
zink, or the entire GLES2.functional.texture.filtering.* set:

dEQP-GLES2.functional.texture.filtering.cube.nearest_linear_clamp_rgba8888_npot
dEQP-GLES2.functional.texture.filtering.cube.nearest_linear_clamp_rgba4444_npot

The problem seems to be zink reinterpreting the flags buffer for a rgba8888
image as the flags buffer for a rgba4444 image, which with UBWC enabled
blows up spectacularly, hanging the entire machine and causing an
insta-reboot. To fix this we need to clear UBWC when the state of the
underlying memory of an image is unknown, which in Vulkan is indicated
by a transition from UNDEFINED. I was holding back on this, even though
freedreno gallium had to do something similar in !4290, because the blob
didn't seem to do it, but now we have conclusive proof that we need it.

In order to properly handle this, we now also have to properly handle
automatic layout transitions in subpasses.
---
 src/freedreno/vulkan/tu_clear_blit.c |  93 ++++++++++++++++
 src/freedreno/vulkan/tu_clear_blit.h |   6 ++
 src/freedreno/vulkan/tu_cmd_buffer.c | 156 +++++++++++++++++++++++++--
 src/freedreno/vulkan/tu_pass.c       |  47 ++++----
 src/freedreno/vulkan/tu_pass.h       |   6 +-
 src/freedreno/vulkan/tu_util.h       |  11 ++
 6 files changed, 283 insertions(+), 36 deletions(-)

diff --git a/src/freedreno/vulkan/tu_clear_blit.c b/src/freedreno/vulkan/tu_clear_blit.c
index 46b53429294d..2f745d6a6d43 100644
--- a/src/freedreno/vulkan/tu_clear_blit.c
+++ b/src/freedreno/vulkan/tu_clear_blit.c
@@ -321,6 +321,30 @@ r2d_dst_buffer(struct tu_cs *cs, enum pipe_format format, uint64_t va, uint32_t
                    A6XX_RB_2D_DST_PITCH(pitch));
 }
 
+/* Write to the flags buffer of a given miplevel/layer as if it's a regular
+ * image.
+ */
+static void
+r2d_dst_ubwc(struct tu_cs *cs,
+             struct tu_image *image,
+             uint32_t plane,
+             uint32_t miplevel,
+             uint32_t layer)
+{
+   const struct fdl_layout *layout = &image->layout[plane];
+   assert(layout->ubwc);
+
+   uint32_t ubwc_offset = fdl_ubwc_offset(layout, miplevel, layer);
+   uint32_t ubwc_pitch = fdl_ubwc_pitch(layout, miplevel);
+   tu_cs_emit_regs(cs,
+                   A6XX_RB_2D_DST_INFO(
+                      .color_format = FMT6_8_UINT,
+                      .tile_mode = TILE6_LINEAR,
+                      .color_swap = WZYX),
+                   A6XX_RB_2D_DST(.qword = image->iova + ubwc_offset),
+                   A6XX_RB_2D_DST_PITCH(ubwc_pitch));
+}
+
 static void
 r2d_setup_common(struct tu_cmd_buffer *cmd,
                  struct tu_cs *cs,
@@ -3417,3 +3441,72 @@ tu_store_gmem_attachment(struct tu_cmd_buffer *cmd,
 
    trace_end_gmem_store(&cmd->trace, cs, dst->format, false, unaligned);
 }
+
+void
+tu_clear_ubwc(struct tu_cmd_buffer *cmd,
+              struct tu_cs *cs,
+              struct tu_image *image,
+              const VkImageSubresourceRange *range)
+{
+   if (!image->layout[0].ubwc)
+      return;
+
+   uint32_t level_count = vk_image_subresource_level_count(&image->vk, range);
+   uint32_t layer_count = vk_image_subresource_layer_count(&image->vk, range);
+
+   r2d_setup(cmd, cs, PIPE_FORMAT_NONE, PIPE_FORMAT_R8_UINT,
+             VK_IMAGE_ASPECT_COLOR_BIT, 0, true, false,
+             VK_SAMPLE_COUNT_1_BIT);
+
+   r2d_clear_value(cs, PIPE_FORMAT_R8_UINT,
+                   &(VkClearValue) { .color = { .uint32 = { 0 }, } });
+
+   unsigned plane_indices[ARRAY_SIZE(image->layout)];
+   unsigned plane_count = 0;
+
+   u_foreach_bit (aspect, range->aspectMask) {
+      unsigned plane_index = tu6_plane_index(image->vk.format, 1u << aspect);
+
+      bool add_index = true;
+      for (unsigned i = 0; i < plane_count; i++) {
+         if (plane_indices[i] == plane_index) {
+            add_index = false;
+            break;
+         }
+      }
+
+      if (add_index) {
+         assert(plane_count < ARRAY_SIZE(plane_indices));
+         plane_indices[plane_count++] = plane_index;
+      }
+   }
+
+   if (image->layout[0].depth0 > 1) {
+      assert(layer_count == 1);
+      assert(range->baseArrayLayer == 0);
+   }
+
+   for (unsigned i = 0; i < plane_count; i++) {
+      unsigned plane = plane_indices[i];
+      if (!image->layout[plane].ubwc)
+         continue;
+
+      for (unsigned j = 0; j < level_count; j++) {
+         unsigned level = j + range->baseMipLevel;
+         if (image->layout[i].depth0 > 1)
+            layer_count = u_minify(image->layout[0].depth0, level);
+
+         r2d_coords(cs, &(VkOffset2D){}, NULL, &(VkExtent2D) {
+                     u_minify(image->layout[i].ubwc_width0, level),
+                     u_minify(image->layout[i].ubwc_height0, level),
+                   });
+
+         for (unsigned k = 0; k < layer_count; k++) {
+            unsigned layer = k + range->baseArrayLayer;
+            r2d_dst_ubwc(cs, image, plane, level, layer);
+            r2d_run(cmd, cs);
+         }
+      }
+   }
+}
+
diff --git a/src/freedreno/vulkan/tu_clear_blit.h b/src/freedreno/vulkan/tu_clear_blit.h
index 77a289afa436..5fed6cf73a12 100644
--- a/src/freedreno/vulkan/tu_clear_blit.h
+++ b/src/freedreno/vulkan/tu_clear_blit.h
@@ -61,4 +61,10 @@ tu_store_gmem_attachment(struct tu_cmd_buffer *cmd,
 void
 tu_choose_gmem_layout(struct tu_cmd_buffer *cmd);
 
+void
+tu_clear_ubwc(struct tu_cmd_buffer *cmd,
+              struct tu_cs *cs,
+              struct tu_image *image,
+              const VkImageSubresourceRange *range);
+
 #endif /* TU_CLEAR_BLIT_H */
diff --git a/src/freedreno/vulkan/tu_cmd_buffer.c b/src/freedreno/vulkan/tu_cmd_buffer.c
index f7c89a2af41c..7260196fdae2 100644
--- a/src/freedreno/vulkan/tu_cmd_buffer.c
+++ b/src/freedreno/vulkan/tu_cmd_buffer.c
@@ -3799,11 +3799,6 @@ tu_subpass_barrier(struct tu_cmd_buffer *cmd_buffer,
    enum tu_cmd_access_mask dst_flags =
       vk2tu_access(barrier->dst_access_mask, dst_stage_vk, false, false);
 
-   if (barrier->incoherent_ccu_color)
-      src_flags |= TU_ACCESS_CCU_COLOR_INCOHERENT_WRITE;
-   if (barrier->incoherent_ccu_depth)
-      src_flags |= TU_ACCESS_CCU_DEPTH_INCOHERENT_WRITE;
-
    tu_flush_for_access(cache, src_flags, dst_flags);
 
    enum tu_stage src_stage = vk2tu_src_stage(src_stage_vk);
@@ -3811,6 +3806,102 @@ tu_subpass_barrier(struct tu_cmd_buffer *cmd_buffer,
    tu_flush_for_stage(cache, src_stage, dst_stage);
 }
 
+/* Handle automatic layout transitions, which at the moment just means
+ * clearing UBWC. Because of this VU of VkAttachmentReference: "If attachment
+ * is not VK_ATTACHMENT_UNUSED, layout must not be VK_IMAGE_LAYOUT_UNDEFINED,
+ * VK_IMAGE_LAYOUT_PREINITIALIZED, or VK_IMAGE_LAYOUT_PRESENT_SRC_KHR", we
+ * don't need to worry about transitions from UNDEFINED except for
+ * the automatic transition from initialLayout, and as long as the attachment
+ * is used then we don't have to worry about which layout it's used with.
+ *
+ * TODO: In sysmem mode, it might be possible to avoid clearing UBWC once we
+ * implement fast clears if we know we'll use a fast clear.
+ *
+ * TODO: In sysmem we need to delay this until first use to handle aliased
+ * attachments. However in gmem we need to keep it like this because we can't
+ * clear UBWC of an attachment in the middle of the loop over tiles.
+ */
+static void
+transition_attachments(struct tu_cmd_buffer *cmd,
+                       struct tu_cs *cs)
+{
+   const struct tu_render_pass *pass = cmd->state.pass;
+   struct tu_cache_state *cache = &cmd->state.cache;
+
+   bool flushed = false;
+   for (unsigned i = 0; i < pass->attachment_count; i++) {
+      if (tu_layout_undefined(pass->attachments[i].initial_layout) &&
+          /* If an attachment is used, we can assume its layout is not
+           * UNDEFINED. If it's not used, then we need to check the final
+           * layout.
+           */
+          (pass->attachments[i].used ||
+           !tu_layout_undefined(pass->attachments[i].final_layout))) {
+         if (!flushed) {
+            tu_flush_for_access(cache, 0,
+                                TU_ACCESS_CCU_COLOR_INCOHERENT_WRITE);
+            /* TODO properly flush for stage */
+            cache->flush_bits |= TU_CMD_FLAG_WAIT_FOR_IDLE;
+            flushed = true;
+         }
+
+         const struct tu_image_view *image_view = cmd->state.attachments[i];
+         if (!image_view->image->layout[0].ubwc)
+            continue;
+
+         /* We're technically inside the renderpass, but we're performing
+          * these flushes before the renderpass actually starts, so we need to
+          * manually flush and set sysmem mode.
+          */
+         tu_emit_cache_flush_ccu(cmd, cs, TU_CMD_CCU_SYSMEM);
+
+         /* From the Vulkan 1.3.207 spec:
+          *
+          *    "Automatic layout transitions apply to the entire image
+          *    subresource attached to the framebuffer. If multiview is not
+          *    enabled and the attachment is a view of a 1D or 2D image, the
+          *    automatic layout transitions apply to the number of layers
+          *    specified by VkFramebufferCreateInfo::layers. If multiview is
+          *    enabled and the attachment is a view of a 1D or 2D image, the
+          *    automatic layout transitions apply to the layers corresponding to
+          *    views which are used by some subpass in the render pass, even if
+          *    that subpass does not reference the given attachment. If the
+          *    attachment view is a 2D or 2D array view of a 3D image, even if
+          *    the attachment view only refers to a subset of the slices of the
+          *    selected mip level of the 3D image, automatic layout transitions
+          *    apply to the entire subresource referenced which is the entire mip
+          *    level in this case."
+          */
+         unsigned view_mask = pass->multiview_mask;
+         VkImageSubresourceRange range = {
+            .aspectMask = image_view->vk.aspects,
+            .baseMipLevel = image_view->vk.base_mip_level,
+            .levelCount = 1,
+         };
+
+         if (image_view->image->vk.image_type == VK_IMAGE_TYPE_3D) {
+            range.baseArrayLayer = 0;
+            range.layerCount = image_view->vk.extent.depth;
+            tu_clear_ubwc(cmd, cs, image_view->image, &range);
+         } else if (view_mask) {
+            range.layerCount = 1;
+            u_foreach_bit (view, view_mask) {
+               range.baseArrayLayer = image_view->vk.base_array_layer + view;
+               tu_clear_ubwc(cmd, cs, image_view->image, &range);
+            }
+         } else {
+            range.baseArrayLayer = image_view->vk.base_array_layer;
+            range.layerCount = cmd->state.framebuffer->layers;
+            tu_clear_ubwc(cmd, cs, image_view->image, &range);
+         }
+      }
+   }
+
+   if (flushed) {
+      tu_flush_for_access(cache, TU_ACCESS_CCU_COLOR_INCOHERENT_WRITE, 0);
+   }
+}
+
 /* emit mrt/zs/msaa/ubwc state for the subpass that is starting (either at
  * vkCmdBeginRenderPass2() or vkCmdNextSubpass2())
  */
@@ -3875,6 +3966,7 @@ tu_CmdBeginRenderPass2(VkCommandBuffer commandBuffer,
     * of the subpass.
     */
    tu_subpass_barrier(cmd, &pass->subpasses[0].start_barrier, true);
+   transition_attachments(cmd, &cmd->cs);
    cmd->state.renderpass_cache.pending_flush_bits =
       cmd->state.cache.pending_flush_bits;
    cmd->state.renderpass_cache.flush_bits = 0;
@@ -5509,12 +5601,60 @@ tu_barrier(struct tu_cmd_buffer *cmd,
       }
    }
 
+   enum tu_stage src_stage = vk2tu_src_stage(srcStage);
+   enum tu_stage dst_stage = vk2tu_dst_stage(dstStage);
+
+   bool flushed = false;
    struct tu_cache_state *cache =
-      cmd->state.pass  ? &cmd->state.renderpass_cache : &cmd->state.cache;
+      cmd->state.pass ? &cmd->state.renderpass_cache : &cmd->state.cache;
+   struct tu_cs *cs = &cmd->cs;
+   for (uint32_t i = 0; i < dep_info->imageMemoryBarrierCount; i++) {
+      VkImageLayout old_layout = dep_info->pImageMemoryBarriers[i].oldLayout;
+      VkImageLayout new_layout = dep_info->pImageMemoryBarriers[i].newLayout;
+      /* The underlying memory for this image may have been used earlier
+       * within the same queue submission for a different image, which
+       * means that there may be old, stale cache entries which are in the
+       * "wrong" location, which could cause problems later after writing
+       * to the image. We don't want these entries being flushed later and
+       * overwriting the actual image, so we need to flush the CCU.
+       * 
+       * We also don't know what state the flags are in, and some "invalid"
+       * states have been observed to produce incorrect rendering or hangs, so
+       * we need to clear the flags buffer for any images with UBWC. All 0 is
+       * guaranteed to be a "valid" state.
+       */
+      if (tu_layout_undefined(old_layout) &&
+          !tu_layout_undefined(new_layout)) {
+         /* From section 7.6.1 "Subpass Self-dependency" of the Vulkan 1.3.231
+          * spec:
+          *
+          *    "Image memory barriers ... must not define an image layout
+          *    transition"
+          *
+          * This means we don't have to worry about layout transitions inside
+          * a subpass, which would be impossible to implement in GMEM mode.
+          */
+         assert(!cmd->state.pass);
+
+         TU_FROM_HANDLE(tu_image, image, dep_info->pImageMemoryBarriers[i].image);
+
+         if (!flushed) {
+            tu_flush_for_access(cache, src_flags,
+                                TU_ACCESS_CCU_COLOR_INCOHERENT_WRITE);
+            tu_flush_for_stage(cache, src_stage, TU_STAGE_PS);
+            src_flags = TU_ACCESS_CCU_COLOR_INCOHERENT_WRITE;
+            src_stage = TU_STAGE_PS;
+
+            flushed = true;
+         }
+
+         tu_clear_ubwc(cmd, cs, image,
+                       &dep_info->pImageMemoryBarriers[i].subresourceRange);
+      }
+   }
+
    tu_flush_for_access(cache, src_flags, dst_flags);
 
-   enum tu_stage src_stage = vk2tu_src_stage(srcStage);
-   enum tu_stage dst_stage = vk2tu_dst_stage(dstStage);
    tu_flush_for_stage(cache, src_stage, dst_stage);
 }
 
diff --git a/src/freedreno/vulkan/tu_pass.c b/src/freedreno/vulkan/tu_pass.c
index 84c1c306129b..5bb97b250251 100644
--- a/src/freedreno/vulkan/tu_pass.c
+++ b/src/freedreno/vulkan/tu_pass.c
@@ -116,19 +116,6 @@ tu_render_pass_add_subpass_dep(struct tu_render_pass *pass,
    dst_barrier->dst_access_mask |= dst_access_mask;
 }
 
-/* We currently only care about undefined layouts, because we have to
- * flush/invalidate CCU for those. PREINITIALIZED is the same thing as
- * UNDEFINED for anything not linear tiled, but we don't know yet whether the
- * images used are tiled, so just assume they are.
- */
-
-static bool
-layout_undefined(VkImageLayout layout)
-{
-   return layout == VK_IMAGE_LAYOUT_UNDEFINED ||
-          layout == VK_IMAGE_LAYOUT_PREINITIALIZED;
-}
-
 /* This implements the following bit of spec text:
  *
  *    If there is no subpass dependency from VK_SUBPASS_EXTERNAL to the
@@ -389,18 +376,9 @@ tu_render_pass_add_implicit_deps(struct tu_render_pass *pass,
       }
    }
 
-   /* Handle UNDEFINED transitions, similar to the handling in tu_barrier().
-    * Assume that if an attachment has an initial layout of UNDEFINED, it gets
-    * transitioned eventually.
-    */
    for (unsigned i = 0; i < info->attachmentCount; i++) {
-      if (layout_undefined(att[i].initialLayout)) {
-         if (vk_format_is_depth_or_stencil(att[i].format)) {
-            pass->subpasses[0].start_barrier.incoherent_ccu_depth = true;
-         } else {
-            pass->subpasses[0].start_barrier.incoherent_ccu_color = true;
-         }
-      }
+      pass->attachments[i].initial_layout = att[i].initialLayout;
+      pass->attachments[i].final_layout = att[i].finalLayout;
    }
 }
 
@@ -759,6 +737,7 @@ tu_subpass_use_attachment(struct tu_render_pass *pass, int i, uint32_t a, const
    struct tu_subpass *subpass = &pass->subpasses[i];
 
    pass->attachments[a].gmem = true;
+   pass->attachments[a].used = true;
    update_samples(subpass, pCreateInfo->pAttachments[a].samples);
    pass->attachments[a].clear_views |= subpass->multiview_mask;
 }
@@ -864,6 +843,7 @@ tu_CreateRenderPass2(VkDevice _device,
       subpass->raster_order_attachment_access = desc->flags & raster_order_access_bits;
 
       subpass->multiview_mask = desc->viewMask;
+      pass->multiview_mask |= subpass->multiview_mask;
 
       if (desc->inputAttachmentCount > 0) {
          subpass->input_attachments = p;
@@ -876,6 +856,9 @@ tu_CreateRenderPass2(VkDevice _device,
              * directly instead of through gmem, so we don't mark input
              * attachments as needing gmem.
              */
+            if (a != VK_ATTACHMENT_UNUSED) {
+               pass->attachments[a].used = true;
+            }
          }
       }
 
@@ -906,8 +889,11 @@ tu_CreateRenderPass2(VkDevice _device,
 
             uint32_t src_a = desc->pColorAttachments[j].attachment;
             if (src_a != VK_ATTACHMENT_UNUSED) {
-               pass->attachments[src_a].will_be_resolved =
-                  desc->pResolveAttachments[j].attachment != VK_ATTACHMENT_UNUSED;
+               uint32_t a = desc->pResolveAttachments[j].attachment;
+               if (a != VK_ATTACHMENT_UNUSED) {
+                  pass->attachments[src_a].will_be_resolved = true;
+                  pass->attachments[a].used = true;
+               }
             }
          }
       }
@@ -920,7 +906,10 @@ tu_CreateRenderPass2(VkDevice _device,
 
          uint32_t src_a = desc->pDepthStencilAttachment->attachment;
          if (src_a != VK_ATTACHMENT_UNUSED) {
-            pass->attachments[src_a].will_be_resolved = a != VK_ATTACHMENT_UNUSED;
+            if (a != VK_ATTACHMENT_UNUSED) {
+               pass->attachments[src_a].will_be_resolved = true;
+               pass->attachments[a].used = true;
+            }
          }
       }
 
@@ -1020,6 +1009,8 @@ tu_setup_dynamic_render_pass(struct tu_cmd_buffer *cmd_buffer,
    subpass->raster_order_attachment_access = false;
    subpass->multiview_mask = info->viewMask;
 
+   pass->multiview_mask = subpass->multiview_mask;
+
    uint32_t a = 0;
    for (uint32_t i = 0; i < info->colorAttachmentCount; i++) {
       struct tu_render_pass_attachment *att = &pass->attachments[a];
@@ -1141,6 +1132,8 @@ tu_setup_dynamic_inheritance(struct tu_cmd_buffer *cmd_buffer,
    subpass->multiview_mask = info->viewMask;
    subpass->samples = info->rasterizationSamples;
 
+   pass->multiview_mask = subpass->multiview_mask;
+
    unsigned a = 0;
    for (unsigned i = 0; i < info->colorAttachmentCount; i++) {
       struct tu_render_pass_attachment *att = &pass->attachments[a];
diff --git a/src/freedreno/vulkan/tu_pass.h b/src/freedreno/vulkan/tu_pass.h
index 06f1185a115f..eb82b36642b4 100644
--- a/src/freedreno/vulkan/tu_pass.h
+++ b/src/freedreno/vulkan/tu_pass.h
@@ -27,7 +27,6 @@ struct tu_subpass_barrier {
    VkPipelineStageFlags2 dst_stage_mask;
    VkAccessFlags2 src_access_mask;
    VkAccessFlags2 dst_access_mask;
-   bool incoherent_ccu_color, incoherent_ccu_depth;
 };
 
 struct tu_subpass_attachment
@@ -73,6 +72,7 @@ struct tu_subpass
 struct tu_render_pass_attachment
 {
    VkFormat format;
+   VkImageLayout initial_layout, final_layout;
    uint32_t samples;
    uint32_t cpp;
    VkImageAspectFlags clear_mask;
@@ -80,6 +80,7 @@ struct tu_render_pass_attachment
    bool load;
    bool store;
    bool gmem;
+   bool used;
    int32_t gmem_offset[TU_GMEM_LAYOUT_COUNT];
    bool will_be_resolved;
    /* for D32S8 separate stencil: */
@@ -106,6 +107,9 @@ struct tu_render_pass
    uint32_t gmem_bandwidth_per_pixel;
    uint32_t sysmem_bandwidth_per_pixel;
 
+   /* logical OR of all views used by all subpasses */
+   uint32_t multiview_mask;
+
    struct tu_subpass_attachment *subpass_attachments;
    struct tu_render_pass_attachment *attachments;
    struct tu_subpass_barrier end_barrier;
diff --git a/src/freedreno/vulkan/tu_util.h b/src/freedreno/vulkan/tu_util.h
index a93b83bd3e75..76e3891a934f 100644
--- a/src/freedreno/vulkan/tu_util.h
+++ b/src/freedreno/vulkan/tu_util.h
@@ -373,6 +373,17 @@ tu6_pack_border_color(struct bcolor_entry *bcolor, const VkClearColorValue *val,
 #undef PACK_F
 }
 
+/* PREINITIALIZED is the same thing as UNDEFINED for anything not linear
+ * tiled, but for now we only care about layouts in the tiled + UBWC case.
+ */
+
+static bool
+tu_layout_undefined(VkImageLayout layout)
+{
+   return layout == VK_IMAGE_LAYOUT_UNDEFINED ||
+          layout == VK_IMAGE_LAYOUT_PREINITIALIZED;
+}
+
 void
 tu_dbg_log_gmem_load_store_skips(struct tu_device *device);
 
-- 
GitLab

